openedx/core/lib/api/authentication.py

```diff
@@ -125,3 +127,110 @@ def _get_dot_token(self, access_token):
+        elif len(auth) > 2:
+            raise AuthenticationFailed({
+                'error_code': OAUTH2_TOKEN_ERROR_MALFORMED,
+                'developer_message': 'Invalid token header. Token string should not contain spaces.'})
+
+        if auth and auth[0].lower() == b'bearer':
+            access_token = auth[1].decode('utf8')
+        elif 'access_token' in request.POST:
+            access_token = request.POST['access_token']
+        elif 'access_token' in request.GET and self.allow_query_params_token:
+            access_token = request.GET['access_token']
+        else:
+            logger.warning("auth is empty")
+            set_custom_metric("OAuth2Authentication", "None")
+            return None
+        user, token = self.authenticate_credentials(access_token)
+        if not self.allow_inactive_users:
+            if not user.is_active:
+                msg = 'User inactive or deleted: %s' % user.get_username()
+                raise AuthenticationFailed(msg)
+
+        set_custom_metric("OAuth2Authentication", "Success")
+
+        return user, token
+
+    def authenticate_credentials(self, access_token):
+        """
+        Authenticate the request, given the access token.
+
+        Overrides base class implementation to discard failure if user is
+        inactive.
+        """
+
+        token = self.get_access_token(access_token)
+        if not token:
+            raise AuthenticationFailed({
+                'error_code': OAUTH2_TOKEN_ERROR_NONEXISTENT,
+                'developer_message': 'The provided access token does not match any valid tokens.'
+            })
+        elif token.expires < django.utils.timezone.now():
+            raise AuthenticationFailed({
+                'error_code': OAUTH2_TOKEN_ERROR_EXPIRED,
+                'developer_message': 'The provided access token has expired and is no longer valid.',
+            })
+        else:
+            return token.user, token
+
+    def get_access_token(self, access_token):
+        """
+        Return a valid access token that exists in one of our OAuth2 libraries,
+        or None if no matching token is found.
+        """
+        return self._get_dot_token(access_token) or self._get_dop_token(access_token)
+
+    def _get_dop_token(self, access_token):
+        """
+        Return a valid access token stored by django-oauth2-provider (DOP), or
+        None if no matching token is found.
+        """
+        token_query = dop_models.AccessToken.objects.select_related('user')
+        return token_query.filter(token=access_token).first()
+
+    def _get_dot_token(self, access_token):
+        """
+        Return a valid access token stored by django-oauth-toolkit (DOT), or
+        None if no matching token is found.
+        """
+        token_query = dot_models.AccessToken.objects.select_related('user')
+        return token_query.filter(token=access_token).first()
```

Comment:
What is https://github.com/jpadilla/django-rest-framework-oauth/blob/master/rest_framework_oauth/authentication.py#L170-L172 ?  Is this important?  Which is more correct for handling expiration?